middlewares/cors.go

```diff
@@ -0,0 +1,100 @@
+package middlewares
+
+import (
+	"github.com/savsgio/atreugo/v10"
+
+	"errors"
+	"strconv"
+	"strings"
+)
+
+// Create new properties for CORS.
+type CorsOptions struct {
+	// given origin or origins can be shared with response
+	AllowedOrigins []string
+
+	// which headers can be used during the actual request
+	AllowedHeaders []string
+
+	// specify the method or methods allowed to resource
+	AllowedMethods []string
+
+	// which header names can be exposed as part of the response
+	ExposedHeaders []string
+
+	// whether a cached response can be used
+	AllowedVary []string
+
+	// seconds for caching the preflight result
+	AllowMaxAge int
+
+	// whether to expose the response to frontend code
+	AllowCredentials bool
+}
+
+type CorsHandler struct {
+	options CorsOptions
+}
+
+func NewCorsMiddleware(options CorsOptions) atreugo.Middleware {
+	cors := CorsHandler{
+		options: options,
+	}
+
+	return cors.middleware
+}
+
+func (c *CorsHandler) middleware(ctx *atreugo.RequestCtx) error {
+	if err := c.handlePreflight(ctx); err != nil {
+		return err
+	}
+
+	return ctx.Next()
+}
+
+func (c *CorsHandler) handlePreflight(ctx *atreugo.RequestCtx) error {
```

Comment:
Why do you change this code again?

```go 
func (c *CorsHandler) handlePreflight(ctx *atreugo.RequestCtx) error {
	originHeader := string(ctx.Request.Header.Peek("Origin"))
	if !c.isAllowedOrigin(originHeader) {
		return ctx.Next()
	}

	// Mandatory Access-Control-Allow-Origin
	ctx.Response.Header.Set("Access-Control-Allow-Origin", originHeader)

	if c.options.AllowCredentials {
		ctx.Response.Header.Set("Access-Control-Allow-Credentials", "true")
	}

	varyHeader := ctx.Response.Header.Peek("Vary")
	if len(varyHeader) > 0 {
		varyHeader = append(varyHeader, ", "...)
	}

	varyHeader = append(varyHeader, "Origin"...)
	ctx.Response.Header.SetBytesV("Vary", varyHeader)

	method := string(ctx.Method())
	if method != fasthttp.MethodOptions {
		return ctx.Next()
	}

	if len(c.options.AllowedHeaders) > 0 {
		ctx.Response.Header.Set("Access-Control-Allow-Headers", strings.Join(c.options.AllowedHeaders, ", "))
	}

	if len(c.options.AllowedMethods) > 0 {
		ctx.Response.Header.Set("Access-Control-Allow-Methods", strings.Join(c.options.AllowedMethods, ", "))
	}

	if len(c.options.ExposedHeaders) > 0 {
		ctx.Response.Header.Set("Access-Control-Expose-Headers", strings.Join(c.options.ExposedHeaders, ", "))
	}

	if c.options.AllowMaxAge > 0 {
		ctx.Response.Header.Set("Access-Control-Max-Age", strconv.Itoa(c.options.AllowMaxAge))
	}

	return ctx.Next()
}
```

Please, change with this code and i will merge it.

Reply:
>Why do you change this code again?

Which code ? func handlePreflight ?

I grouped other similar lines

If you mean these
```
	varyHeader := ctx.Response.Header.Peek("Vary")
	if len(varyHeader) > 0 {
		varyHeader = append(varyHeader, ", "...)
	}

	varyHeader = append(varyHeader, "Origin"...)
	ctx.Response.Header.SetBytesV("Vary", varyHeader)

	method := string(ctx.Method())
	if method != fasthttp.MethodOptions {
		return ctx.Next()
	}
```
1.
	`method != fasthttp.MethodOptions` will allways false cause of fasthttp.MethodOption is OPTIONS every time.

2.
  `append(varyHeader, ", "...)` more vary from request,  why not limited safely?
Is { Vary: "Origin" } only required in CORS ?

Reply:
3.
 ```
	if !c.isAllowedOrigin(originHeader) {
		return ctx.Next()
	}
```
return ctx.Next() allways nil, How to write this test 

``
	if err := c.handlePreflight(ctx); err != nil {
		return err
	}
``